test/functional/wallet_avoid_mixing_output_types.py

```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test output type mixing during coin selection
+
+A wallet may have different type UTXOs to choose from during coin selection,
+where output type is one of the following:
+    - BECH32M
+    - BECH32
+    - P2SH-SEGWIT
+    - LEGACY
+
+This test verifies that mixing different output types is avoided unless
+absolutely necessary. Both wallets start with zero funds. Alice mines
+enough blocks to have spendable coinbase outputs. Alice sends 10
+random value payments to Bob, creating 5 bech32 UTXOs and 5 P2SH UTXOs,
+for a total of 30BTC in Bob's wallet.
+
+Bob then sends random valued payments back to Alice, some of which need
+unconfirmed change, and we verify that none of these payments contain mixed
+inputs. Finally, Bob sends the remainder of his funds, which requires mixing.
+
+The payment values are random, but chosen such that they sum up to a specified
+total. This ensures we are not relying on specific values for the UTXOs,
+but still know when to expect mixing due to the wallet being close to empty.
+
+"""
+
+import random
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.blocktools import COINBASE_MATURITY
+
+ADDRESS_TYPES = [
+    "bech32m",
+    "bech32",
+    "p2sh-segwit",
+    "legacy",
+]
+
+
+def is_bech32_address(node, addr):
+    """Check if an address contains a bech32 output."""
+    addr_info = node.getaddressinfo(addr)
+    if addr_info['iswitness']:
+        return (
+            not addr_info['isscript'] and
+            addr_info['iswitness'] and addr_info['witness_version'] == 0
+        )
+    return False
+
+
+def is_bech32m_address(node, addr):
+    """Check if an address contains a bech32m output."""
+    addr_info = node.getaddressinfo(addr)
+    if addr_info['iswitness']:
+        return (
+            addr_info['isscript'] and
+            addr_info['iswitness'] and addr_info['witness_version'] == 1
+        )
+    return False
+
+
+def is_p2sh_segwit_address(node, addr):
+    """Check if an address contains a P2SH-Segwit outputself.
+       Note: this function does not actually determine the type
+       of P2SH output, but is sufficient for this test in that
+       we are only generating P2SH-Segwit outputs.
+    """
+    addr_info = node.getaddressinfo(addr)
+    return (
+        addr_info['isscript'] and
+        not addr_info['iswitness']
+    )
```

Comment:
In ffcc600156e45968b47586e6758f107a56d6b80e "test: functional test for new coin selection logic"

The easiest way to check for address type is to check what the descriptor starts with. `sh(wpkh(` or `sh(wsh(` is p2sh-segwit, `wpkh(` or `wsh(` is bech32, `tr(` is bech32m, and `pkh(` is legacy.

Reply:
nice! this test only runs with `--descriptors` so ill update the `is_<addrtype>` functions